ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,281 @@
+    "The number of bytes sent/received using RDMA Put operations.",
+    "The number of bytes sent/received using RDMA Get operations.",
+    "The number of messages that arrived as unexpected messages.",
+    "The number of messages that arrived out of the proper sequence.",
+    "The number of microseconds spent matching unexpected messages.",
+    "The number of microseconds spent matching out of sequence messages."
+};
+
+/* An array of integer values to denote whether an event is activated (1) or not (0) */
+OMPI_DECLSPEC unsigned int attached_event[OMPI_NUM_COUNTERS] = { 0 };
+/* An array of integer values to denote whether an event is timer-based (1) or not (0) */
+OMPI_DECLSPEC unsigned int timer_event[OMPI_NUM_COUNTERS] = { 0 };
+/* An array of event structures to store the event data (name and value) */
+OMPI_DECLSPEC ompi_event_t *events = NULL;
+
+/* ##############################################################
+ * ################# Begin MPI_T Functions ######################
+ * ##############################################################
+ */
+static int ompi_spc_notify(mca_base_pvar_t *pvar, mca_base_pvar_event_t event, void *obj_handle, int *count)
+{
+    (void)obj_handle;
+
+    int i;
+
+    /* For this event, we need to set count to the number of long long type
+     * values for this counter.  All SPC counters are one long long, so we
+     * always set count to 1.
+     */
+    if(MCA_BASE_PVAR_HANDLE_BIND == event)
+        *count = 1;
+    /* For this event, we need to turn on the counter */
+    else if(MCA_BASE_PVAR_HANDLE_START == event) {
+        /* Loop over the mpi_t_inddices array and find the correct SPC index to turn on */
+        for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+            if(pvar->pvar_index == mpi_t_indices[i]) {
+                attached_event[i] = 1;
+                break;
+            }
+        }
+    }
+    /* For this event, we need to turn off the counter */
+    else if(MCA_BASE_PVAR_HANDLE_STOP == event) {
+        /* Loop over the mpi_t_inddices array and find the correct SPC index to turn off */
+        for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+            if(pvar->pvar_index == mpi_t_indices[i]) {
+                attached_event[i] = 0;
+                break;
+            }
+        }
+    }
+
+    return MPI_SUCCESS;
+}
+
+/* ##############################################################
+ * ################# Begin SPC Functions ########################
+ * ##############################################################
+ */
+
+/* This function returns the current count of an SPC counter that has been retistered
+ * as an MPI_T pvar.  The MPI_T index is not necessarily the same as the SPC index,
+ * so we need to convert from MPI_T index to SPC index and then set the 'value' argument
+ * to the correct value for this pvar.
+ */
+static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, void *obj_handle)
```

Comment:
The read operation is supposed to be fast and constant in time independent on the number of events to look for. Can we have a translation table instead of looping around all counters to find the corresponding SPC index ?

Reply:
So, the MPI_T indices are basically the same as the SPC indices with an offset.  I could potentially save this offset, but I'm not sure this is safe.  Everything in MPI_Init should be serial, right?  If so, this method would be safe since we can assume that the indices will be contiguous.

Reply:
This could also be problematic if any of the SPCs fail to be registered properly with MPI_T.  I'm not sure how I would implement a proper translation table through.

Reply:
If something fails to get registered properly you can just disable the entire component. If you can validate the offset idea that would be really great. In the worst case we might want to reach out to Nathan to see if he considers this as a safe solution.